src/protocol/context/semi_honest.rs

```diff
@@ -1,44 +1,134 @@
+//! Context for protocol executions suitable for semi-honest security model, i.e. secure against
+//! honest-but-curious adversary parties.
+
 use crate::ff::Field;
 use crate::helpers::messaging::{Gateway, Mesh, TotalRecords};
 use crate::helpers::Role;
 use crate::protocol::context::{
     Context, InstrumentedIndexedSharedRandomness, InstrumentedSequentialSharedRandomness,
-    MaliciousContext,
+    MaliciousContext, MaliciousContextBuf,
 };
 use crate::protocol::malicious::MaliciousValidatorAccumulator;
 use crate::protocol::prss::Endpoint as PrssEndpoint;
 use crate::protocol::{Step, Substep};
 use crate::secret_sharing::replicated::semi_honest::AdditiveShare as Replicated;
-use crate::sync::Arc;
 
+use once_cell::sync::OnceCell;
 use std::marker::PhantomData;
 
-/// Context for protocol executions suitable for semi-honest security model, i.e. secure against
-/// honest-but-curious adversary parties.
-#[derive(Clone, Debug)]
-pub struct SemiHonestContext<'a, F: Field> {
-    /// TODO (alex): Arc is required here because of the `TestWorld` structure. Real world
-    /// may operate with raw references and be more efficient
-    pub(super) inner: Arc<ContextInner<'a>>,
+#[derive(Debug)]
+pub struct ContextStore<'a> {
+    pub(super) static_data: ContextStatic<'a>,
+    // Ensure that the struct is invariant over 'a, because that may be required in the future.
+    pub(super) _dummy: PhantomData<OnceCell<&'a ()>>,
+}
+
+impl<'a> ContextStore<'a> {
+    fn new(prss: &'a PrssEndpoint, gateway: &'a Gateway) -> ContextStore<'a> {
+        ContextStore {
+            static_data: ContextStatic::new(prss, gateway),
+            _dummy: PhantomData,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SemiHonestContext<'c, 'a, F: Field> {
+    pub(super) store: &'c ContextStore<'a>,
     pub(super) step: Step,
     pub(super) total_records: TotalRecords,
+    // Ensure that the struct is invariant over 'a, because that may be required in the future.
+    _dummy: PhantomData<OnceCell<&'a ()>>,
     _marker: PhantomData<F>,
 }
 
-impl<'a, F: Field> SemiHonestContext<'a, F> {
-    pub fn new(participant: &'a PrssEndpoint, gateway: &'a Gateway) -> Self {
-        Self::new_with_total_records(participant, gateway, TotalRecords::Unspecified)
-    }
+#[derive(Debug)]
+pub struct SemiHonestContextBuf<'a, F: Field> {
+    pub(super) store: ContextStore<'a>,
+    pub(super) step: Step,
+    pub(super) total_records: TotalRecords,
+    // Ensure that the struct is invariant over 'a, because that may be required in the future.
+    _dummy: PhantomData<OnceCell<&'a ()>>,
+    _marker: PhantomData<F>,
```

Comment:
It might be better to combine these:

```suggestion
    _marker: PhantomData<(&'a (), F)>,
```

Reply:
My thought was that `_dummy` goes away once we add some other member that makes use of interior mutability, whereas `_marker` isn't going anywhere, but I do agree that it otherwise makes sense to be less verbose by combining the two `PhantomData`s.